base/compiler/ssair/inlining.jl

```diff
@@ -1188,16 +1188,23 @@ function handle_invoke_call!(todo::Vector{Pair{Int,Any}},
         item = concrete_result_item(result, info, state; invokesig)
     else
         argtypes = invoke_rewrite(sig.argtypes)
-        if isa(result, ConstPropResult)
-            mi = result.result.linfo
-            validate_sparams(mi.sparam_vals) || return nothing
-            if Union{} !== argtypes_to_type(argtypes) <: mi.def.sig
-                item = resolve_todo(mi, result.result, argtypes, info, flag, state; invokesig)
-                handle_single_case!(todo, ir, idx, stmt, item, true)
-                return nothing
+        if isa(result, SemiConcreteResult)
+            result = inlining_policy(state.interp, result, info, flag, result.mi, argtypes)
+        end
```

Comment:
I remember that we decided to make `SemiConcreteResult` configured by `inlining_policy` to allow external `AbstractInterpreter` to implement a customized inlining policy for semi-concrete eval calls. But I looked around the custom `inlining_policy` implementations and all of them just return input `SemiConcreteResult` as is, so it looks like this unnecessarily complicates the inlining code?

Reply:
Diffractor uses the `info` field to reject inlining requests that have an frule. It is true that it passes through the SemiConcreteResult if allowable, but it checks info first.

Reply:
Right, we need to go through `inlining_policy` when handling `SemiConcreteResult`. I'd argue that it might be clearer if we pass `result.ir` to `inlining_policy` instead of `result::SemiConcreteResult`, since it feels a bit strange that the `src` argument of `inlining_policy` needs to handle `SemiConcreteResult` while it doesn't need to handle the other cache container objects like `CodeInstance` or `InferenceResult`.